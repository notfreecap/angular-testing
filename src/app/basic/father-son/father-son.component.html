<h5>Hijo</h5>
<code>{{ client | json }}</code>
<br>
<button 
    class="btn btn-outline-primary" 
    *ngIf="client" 
    (click)="onChange(5)"
    data-test="btn-change-id">
    Cambiar Id
</button>

<button 
    class="btn btn-outline-danger" 
    *ngIf="client" 
    (click)="onDelete()"
    data-test="btn-delete">
    Borrar cliente
</button>