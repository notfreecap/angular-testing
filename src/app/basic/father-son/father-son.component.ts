import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css']
})
export class FatherSonComponent implements OnInit {

  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  constructor() { }

  ngOnInit(): void {
  }

  onDelete(){
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number){
    if (!this.client) return;

    // this.client.id = newId;
    this.client = {...this.client, id};

    this.onClientUpdated.emit({ ...this.client });
  }

}
