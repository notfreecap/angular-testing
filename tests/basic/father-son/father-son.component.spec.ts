import { createPlatform } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    // your code here
    compiled = fixture.nativeElement;
    // clar all mocks
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe de hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('no deben aparecer bonotes', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('deben aparecer bonotes si hay cliente', () => {
    component.client = {id: 1, name: 'ghost'};
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('debe emitir onDeleteClient con el boton eliminar', () => {
    component.client = {id: 1, name: 'ghost'};
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    // console.log(btnDelete?.textContent);
    btnDelete?.dispatchEvent(new Event('click'));

    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('debe emitir onClientUpdated con el boton "Cambiar ID"', () => {
    component.client = {id: 1, name: 'ghost'};
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');

    const btnChangeId = compiled.querySelector('[data-test=btn-change-id]');
    // console.log(btnDelete?.textContent);
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'ghost'
    });
  });

  test('debe emitir onChangeClient con el ID especificado si hay un cliente', () => {
    jest.spyOn(component.onClientUpdated, 'emit');

    component.onChange(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();

    component.client = {id: 1, name: 'ghost'};
    fixture.detectChanges();
    component.onChange(10);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'ghost'
    });

  });

});
