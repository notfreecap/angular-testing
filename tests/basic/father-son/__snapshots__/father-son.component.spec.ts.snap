// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  
  
</div>
`;
